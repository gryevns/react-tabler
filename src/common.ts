export type Color =
    | "blue"
    | "azure"
    | "indigo"
    | "purple"
    | "pink"
    | "red"
    | "orange"
    | "yellow"
    | "lime"
    | "green"
    | "teal"
    | "cyan"
    | "gray";

export const colors = [
    "blue" as const,
    "azure" as const,
    "indigo" as const,
    "purple" as const,
    "pink" as const,
    "red" as const,
    "orange" as const,
    "yellow" as const,
    "lime" as const,
    "green" as const,
    "teal" as const,
    "cyan" as const,
    "gray" as const,
];

export type Variant =
    | "primary"
    | "secondary"
    | "success"
    | "info"
    | "warning"
    | "danger";

export const variants = [
    "primary" as const,
    "secondary" as const,
    "success" as const,
    "info" as const,
    "warning" as const,
    "danger" as const,
];

export type Size = "sm" | "md" | "lg" | "xl";

export const sizes = ["sm" as const, "md" as const, "lg" as const, "xl" as const];
