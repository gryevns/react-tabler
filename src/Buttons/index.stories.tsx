import React from "react";
import { Link, MemoryRouter as Router } from "react-router-dom";
import { Facebook, Instagram, Twitter } from "react-feather";
import { action } from "@storybook/addon-actions";

import Buttons from ".";
import Tooltip from "../Tooltip";
import { Upload } from "react-feather";
import { colors, Color, variants, Variant } from "../common";

export default {
    title: "Buttons",
};

export const ButtonHTML = () => {
    return (
        <Buttons.Button variant="primary" onClick={action("button-click")}>
            Button
        </Buttons.Button>
    );
};

export const AnchorHTML = () => {
    return (
        <Buttons.Anchor variant="secondary" onClick={action("button-click")}>
            Anchor
        </Buttons.Anchor>
    );
};

export const InputHTML = () => {
    return (
        <Buttons.Input
            variant="warning"
            onClick={action("button-click")}
            value="Input"
        />
    );
};

export const ReactRouterLink = () => (
    <Router>
        <Link to="/" component={Buttons.Button} variant="primary">
            Home
        </Link>
    </Router>
);

export const Variants = () => (
    <Buttons.List>
        {[...variants, "light", "dark", "link"].map((variant: Variant) => (
            <Buttons.Button variant={variant}>{variant}</Buttons.Button>
        ))}
    </Buttons.List>
);

export const Ghost = () => (
    <Buttons.List>
        {[...variants, "light", "dark", "link"].map((variant: Variant) => (
            <Buttons.Button variant={variant} ghost>
                {variant}
            </Buttons.Button>
        ))}
    </Buttons.List>
);

export const Colors = () => (
    <Buttons.List>
        {colors.map((color: Color) => (
            <Buttons.Button color={color}>{color}</Buttons.Button>
        ))}
    </Buttons.List>
);

export const Disabled = () => (
    <Buttons.List>
        {[...variants, "light", "dark", "link"].map((variant: Variant) => (
            <Buttons.Button disabled variant={variant}>
                {variant}
            </Buttons.Button>
        ))}
    </Buttons.List>
);

export const Square = () => (
    <Buttons.List>
        {[...variants, "light", "dark", "link"].map((variant: Variant) => (
            <Buttons.Button square variant={variant}>
                {variant}
            </Buttons.Button>
        ))}
    </Buttons.List>
);

export const Pill = () => (
    <Buttons.List>
        {[...variants, "light", "dark", "link"].map((variant: Variant) => (
            <Buttons.Button pill variant={variant}>
                {variant}
            </Buttons.Button>
        ))}
    </Buttons.List>
);

export const Outline = () => (
    <Buttons.List>
        {[...variants, "light", "dark", "link"].map((variant: Variant) => (
            <Buttons.Button outline variant={variant}>
                {variant}
            </Buttons.Button>
        ))}
    </Buttons.List>
);

export const Loading = () => (
    <Buttons.List>
        {[...variants, "light", "dark", "link"].map((variant: Variant) => (
            <Buttons.Button loading variant={variant}>
                {variant}
            </Buttons.Button>
        ))}
    </Buttons.List>
);

export const Sizes = () =>
    ["sm" as const, "md" as const, "lg" as const].map(size => (
        <Buttons.Button size={size} block={true} variant="primary">
            {size}
        </Buttons.Button>
    ));

export const Tooltips = () => (
    <Tooltip placement="bottom" content="tooltip">
        <Buttons.Button variant="primary">Hover me</Buttons.Button>
    </Tooltip>
);

export const Icons = () => (
    <Buttons.List>
        <Buttons.Anchor variant="primary" icon social="facebook">
            <Facebook size={18} color="white" className="icon" />
        </Buttons.Anchor>
        <Buttons.Anchor variant="primary" icon social="instagram">
            <Instagram size={18} color="white" className="icon" />
        </Buttons.Anchor>
        <Buttons.Anchor variant="primary" icon social="twitter">
            <Twitter size={18} color="white" className="icon" />
        </Buttons.Anchor>
        <Buttons.Button link>
            <Upload size={18} className="icon" />
            &nbsp;Upload
        </Buttons.Button>
    </Buttons.List>
);
