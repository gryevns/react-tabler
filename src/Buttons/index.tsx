import React from "react";
import cn from "classnames";
import { Color, Variant } from "../common";

export interface ListProps {
    className?: string;
    children: React.ReactNode;
}

export const List = ({ className, children }: ListProps) => {
    const classes = cn(`btn-list`, className);
    return <div className={classes}>{children}</div>;
};

export type Size = "sm" | "md" | "lg";

interface ButtonProps {
    className?: string;
    children?: React.ReactNode;
    size?: Size;
    outline?: boolean;
    link?: boolean;
    icon?: boolean;
    block?: boolean;
    disabled?: boolean;
    color?: Color;
    variant?: Variant | "link";
    square?: boolean;
    pill?: boolean;
    loading?: boolean;
    social?: string;
}

const getClassName = ({
    size,
    block,
    variant,
    outline,
    link,
    disabled,
    color,
    pill,
    icon,
    loading,
    square,
    social,
    className,
}: ButtonProps) => {
    const classes = cn(
        {
            btn: true,
            [`btn-${size}`]: !!size,
            [`btn-block`]: block,
            [`btn-outline-${variant}`]: outline && !!variant,
            [`btn-link`]: link,
            disabled: disabled,
            [`btn-${social}`]: !!social,
            [`btn-${color}`]: !!color && !outline,
            [`btn-${variant}`]: !!variant && !outline,
            "btn-square": square,
            "btn-pill": pill,
            "btn-icon": icon,
            "btn-loading": loading,
        },
        className,
    );
    return classes;
};

export const Anchor = React.forwardRef(
    (
        props: ButtonProps & React.AnchorHTMLAttributes<HTMLElement>,
        ref: React.Ref<HTMLAnchorElement>,
    ) => {
        const classes = getClassName(props);
        return (
            <a ref={ref} {...props} className={classes}>
                {props.children}
            </a>
        );
    },
);

export const Input = React.forwardRef(
    (
        props: ButtonProps & React.InputHTMLAttributes<HTMLElement>,
        ref: React.Ref<HTMLInputElement>,
    ) => {
        const classes = getClassName(props);
        return <input type="button" ref={ref} {...props} className={classes} />;
    },
);

export const Button = React.forwardRef(
    (
        props: ButtonProps & React.ButtonHTMLAttributes<HTMLElement>,
        ref: React.Ref<HTMLButtonElement>,
    ) => {
        const classes = getClassName(props);
        return (
            <button ref={ref} {...props} className={classes}>
                {props.children}
            </button>
        );
    },
);

const Buttons = () => {};

Buttons.Anchor = Anchor;
Buttons.Button = Button;
Buttons.Input = Input;
Buttons.List = List;

export default Buttons;
