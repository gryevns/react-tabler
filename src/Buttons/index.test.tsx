import React from "react";
import { shallow } from "enzyme";

import { colors, variants } from "../common";
import Buttons from "./index";

describe("Buttons", () => {
    it("should match button snapshot", () => {
        const wrapper = shallow(<Buttons.Button>Test</Buttons.Button>);
        expect(wrapper).toMatchSnapshot();
    });
    it("should handle onClick for button", () => {
        const mock = jest.fn();
        const wrapper = shallow(
            <Buttons.Button onClick={mock}>Test</Buttons.Button>,
        );
        wrapper.find("button").simulate("click");
        expect(mock).toHaveBeenCalled();
    });
    it("should match anchor snapshot", () => {
        const wrapper = shallow(<Buttons.Anchor>Test</Buttons.Anchor>);
        expect(wrapper).toMatchSnapshot();
    });
    it("should handle onClick for anchor", () => {
        const mock = jest.fn();
        const wrapper = shallow(
            <Buttons.Anchor onClick={mock}>Test</Buttons.Anchor>,
        );
        wrapper.find("a").simulate("click");
        expect(mock).toHaveBeenCalled();
    });
    it("should match input snapshot", () => {
        const wrapper = shallow(<Buttons.Input>Test</Buttons.Input>);
        expect(wrapper).toMatchSnapshot();
    });
    it("should handle onClick for input", () => {
        const mock = jest.fn();
        const wrapper = shallow(<Buttons.Input onClick={mock}>Test</Buttons.Input>);
        wrapper.find("input").simulate("click");
        expect(mock).toHaveBeenCalled();
    });
    it("should match variant snapshot", () => {
        [...variants, "light" as const, "dark" as const, "link" as const].map(
            variant => {
                const wrapper = shallow(
                    <Buttons.Button variant={variant}>{variant}</Buttons.Button>,
                );
                expect(wrapper).toMatchSnapshot();
            },
        );
    });
    it("should match color snaphot", () => {
        colors.map(color => {
            const wrapper = shallow(
                <Buttons.Button color={color}>{color}</Buttons.Button>,
            );
            expect(wrapper).toMatchSnapshot();
        });
    });
    it("should match ghost snapshot", () => {
        const wrapper = shallow(<Buttons.Button ghost>Test</Buttons.Button>);
        expect(wrapper).toMatchSnapshot();
    });
    it("should match disabled snapshot", () => {
        const wrapper = shallow(<Buttons.Button disabled>Test</Buttons.Button>);
        expect(wrapper).toMatchSnapshot();
    });
    it("should match square snapshot", () => {
        const wrapper = shallow(<Buttons.Button square>Test</Buttons.Button>);
        expect(wrapper).toMatchSnapshot();
    });
    it("should match pill snapshot", () => {
        const wrapper = shallow(<Buttons.Button pill>Test</Buttons.Button>);
        expect(wrapper).toMatchSnapshot();
    });
    it("should match outline snapshot", () => {
        const wrapper = shallow(<Buttons.Button outline>Test</Buttons.Button>);
        expect(wrapper).toMatchSnapshot();
    });
    it("should match loading snapshot", () => {
        const wrapper = shallow(<Buttons.Button loading>Test</Buttons.Button>);
        expect(wrapper).toMatchSnapshot();
    });
    it("should match sizes snapshot", () => {
        ["sm" as const, "md" as const, "lg" as const].map(size => {
            const wrapper = shallow(
                <Buttons.Button size={size}>{size}</Buttons.Button>,
            );
            expect(wrapper).toMatchSnapshot();
        });
    });
    it("should match icon snapshot", () => {
        const wrapper = shallow(<Buttons.Button icon>Icon</Buttons.Button>);
        expect(wrapper).toMatchSnapshot();
    });
});
