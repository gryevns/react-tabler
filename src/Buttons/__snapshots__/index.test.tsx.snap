// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Buttons should match anchor snapshot 1`] = `
<a
  className="btn"
>
  Test
</a>
`;

exports[`Buttons should match button snapshot 1`] = `
<button
  className="btn"
>
  Test
</button>
`;

exports[`Buttons should match color snaphot 1`] = `
<button
  className="btn btn-blue"
  color="blue"
>
  blue
</button>
`;

exports[`Buttons should match color snaphot 2`] = `
<button
  className="btn btn-azure"
  color="azure"
>
  azure
</button>
`;

exports[`Buttons should match color snaphot 3`] = `
<button
  className="btn btn-indigo"
  color="indigo"
>
  indigo
</button>
`;

exports[`Buttons should match color snaphot 4`] = `
<button
  className="btn btn-purple"
  color="purple"
>
  purple
</button>
`;

exports[`Buttons should match color snaphot 5`] = `
<button
  className="btn btn-pink"
  color="pink"
>
  pink
</button>
`;

exports[`Buttons should match color snaphot 6`] = `
<button
  className="btn btn-red"
  color="red"
>
  red
</button>
`;

exports[`Buttons should match color snaphot 7`] = `
<button
  className="btn btn-orange"
  color="orange"
>
  orange
</button>
`;

exports[`Buttons should match color snaphot 8`] = `
<button
  className="btn btn-yellow"
  color="yellow"
>
  yellow
</button>
`;

exports[`Buttons should match color snaphot 9`] = `
<button
  className="btn btn-lime"
  color="lime"
>
  lime
</button>
`;

exports[`Buttons should match color snaphot 10`] = `
<button
  className="btn btn-green"
  color="green"
>
  green
</button>
`;

exports[`Buttons should match color snaphot 11`] = `
<button
  className="btn btn-teal"
  color="teal"
>
  teal
</button>
`;

exports[`Buttons should match color snaphot 12`] = `
<button
  className="btn btn-cyan"
  color="cyan"
>
  cyan
</button>
`;

exports[`Buttons should match color snaphot 13`] = `
<button
  className="btn btn-gray"
  color="gray"
>
  gray
</button>
`;

exports[`Buttons should match disabled snapshot 1`] = `
<button
  className="btn disabled"
  disabled={true}
>
  Test
</button>
`;

exports[`Buttons should match ghost snapshot 1`] = `
<button
  className="btn btn-ghost-undefined"
  ghost={true}
>
  Test
</button>
`;

exports[`Buttons should match icon snapshot 1`] = `
<button
  className="btn btn-icon"
  icon={true}
>
  Icon
</button>
`;

exports[`Buttons should match input snapshot 1`] = `
<input
  className="btn"
  type="button"
>
  Test
</input>
`;

exports[`Buttons should match loading snapshot 1`] = `
<button
  className="btn btn-loading"
  loading={true}
>
  Test
</button>
`;

exports[`Buttons should match outline snapshot 1`] = `
<button
  className="btn"
  outline={true}
>
  Test
</button>
`;

exports[`Buttons should match pill snapshot 1`] = `
<button
  className="btn btn-pill"
  pill={true}
>
  Test
</button>
`;

exports[`Buttons should match sizes snapshot 1`] = `
<button
  className="btn btn-sm"
  size="sm"
>
  sm
</button>
`;

exports[`Buttons should match sizes snapshot 2`] = `
<button
  className="btn btn-md"
  size="md"
>
  md
</button>
`;

exports[`Buttons should match sizes snapshot 3`] = `
<button
  className="btn btn-lg"
  size="lg"
>
  lg
</button>
`;

exports[`Buttons should match square snapshot 1`] = `
<button
  className="btn btn-square"
  square={true}
>
  Test
</button>
`;

exports[`Buttons should match variant snapshot 1`] = `
<button
  className="btn btn-primary"
  variant="primary"
>
  primary
</button>
`;

exports[`Buttons should match variant snapshot 2`] = `
<button
  className="btn btn-secondary"
  variant="secondary"
>
  secondary
</button>
`;

exports[`Buttons should match variant snapshot 3`] = `
<button
  className="btn btn-success"
  variant="success"
>
  success
</button>
`;

exports[`Buttons should match variant snapshot 4`] = `
<button
  className="btn btn-info"
  variant="info"
>
  info
</button>
`;

exports[`Buttons should match variant snapshot 5`] = `
<button
  className="btn btn-warning"
  variant="warning"
>
  warning
</button>
`;

exports[`Buttons should match variant snapshot 6`] = `
<button
  className="btn btn-danger"
  variant="danger"
>
  danger
</button>
`;

exports[`Buttons should match variant snapshot 7`] = `
<button
  className="btn btn-light"
  variant="light"
>
  light
</button>
`;

exports[`Buttons should match variant snapshot 8`] = `
<button
  className="btn btn-dark"
  variant="dark"
>
  dark
</button>
`;

exports[`Buttons should match variant snapshot 9`] = `
<button
  className="btn btn-link"
  variant="link"
>
  link
</button>
`;
